### PR TITLE
http: making routing cookies HttpOnly, allowing configured paths

### LINKED ISSUE #3532: Routing cookie enhancements
When issuing routing cookies, we may want to limit the cookie to the paths specified in the route match.

Also given I can't think of a good reason for them to be accessible from Javascript I think we should probably flag them HttpOnly by default.

### PR DESCRIPTION
**Pull request by alyssawilk**
*Risk Level*: Low
*Testing*: new unit tests
*Docs Changes*:  config documented inline.
*Release Notes*: added in #3462
Fixed #3532

### CODE REVIEW

**Comment by akonradi** [comment]
Just curious, what's the rationale for having a parameter to control adding HttpOnly if the cookie is only for consumption by Envoy? Is this something we'd ever want to be able to turn off?

**Comment by alyssawilk** [comment]
AFIK "HttpOnly" basically blocks the cookie being accessed by client side javascript.  I don't think there's a good reason to allow browsers to mess with Envoy routing - our prior (internal) cookie logic is HttpOnly and I figured we should do the same in Envoy.  If anyone ever wants to make them accessible we can always add a config option for it and let folks opt into client side cookie changes.

File: source/common/http/utility.cc [hunk 1]
```diff
@@ -134,11 +134,13 @@ std::string Utility::parseCookieValue(const HeaderMap& headers, const std::strin
...
 
 std::string Utility::makeSetCookieValue(const std::string& key, const std::string& value,
-                                        const std::chrono::seconds max_age) {
-  if (max_age == std::chrono::seconds::zero()) {
-    return fmt::format("{}=\"{}\"", key, value);
-  }
-  return fmt::format("{}=\"{}\"; Max-Age={}", key, value, max_age.count());
+                                        const std::string& path, const std::chrono::seconds max_age,
+                                        bool httponly) {
+  return fmt::format("{}=\"{}\"{}{}{}{}", key, value,
```
Comment (akonradi) [inline source/common/http/utility.cc]: This is starting to get pretty convoluted, to the point where it's difficult to understand what the output would actually look like. Can you refactor to get rid of the nested `fmt::format`s?

Comment (alyssawilk) [reply source/common/http/utility.cc]: Done, though I'll want a !Googler's take on using absl over fmt here.  Worst case we can do a bunch of string.appends

Oh, I also meant to call out the fact we're doing name="value" but path=/foo.  I'm inclined to take the quotes out around value but wanted to get your thoughts there - if you prefer them in both places or prefer them as-is that's fine too.

Comment (akonradi) [reply source/common/http/utility.cc]: Thanks, that looks a lot better with the StrAppend, but I confess I'm biased.

The rationale for quoting the value is that it gives the caller more flexibility. The HTTP spec disallows backslashes and double quotes in values to begin with, but this allows the caller to set a cookie that includes whitespace. Looking [at the Mozilla docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie) I don't think quoting the path is allowed.

**Comment by PiotrSikora** [comment]
Could you also add `Secure` flag for routing cookies sent over TLS connections? It's similar concept to `HttpOnly`, and should be pretty much the same code change, so it might make sense to include it in this PR (but feel free to skip it).

**Comment by alyssawilk** [comment]
I think Secure should probably be a configurable option - some folks may want the same routing hash on both secure and insecure paths, especially as there's really no PII for these cookies.  Given that, I'm inclined to wait until someone wants it (though as you say it should be a pretty easy addition)

**Pull request merged**